### PR TITLE
Dataclass

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**
This PR changes the base class of `evaluable.Evaluable` to `types.DataClass`, which is a newly introduced base class akin to Python's built in dataclass but implemented as a singleton.

### CODE REVIEW

File: nutils/evaluable.py [hunk 1]
```diff
@@ -134,6 +134,10 @@ class ExpensiveEvaluationWarning(warnings.NutilsInefficiencyWarning):
     pass
 
 
+def _abstract_property(self):
+    raise NotImplementedError(f'{self.__class__.__name__} fails to implement dependencies')
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:138]: The error message refers to some property `dependencies`, but `_abstract_property` is used from `shape` and `dtype`.

File: nutils/types.py [hunk 2]
```diff
@@ -290,6 +290,78 @@ class Singleton(Immutable, metaclass=SingletonMeta):
...
+            self.__post_init__()
+            cls.__cache[bound.args] = self
+        return self
+
+
+class DataClass(metaclass=DataClassMeta):
+    '''
+    Base class for dataclass-like types, for which an assignment only
+    constructor is inferred from type annotations. Post processing is possible
+    via an optional __post_init__ method. The class adds equality tests,
```
Comment (joostvanzwieten) [inline nutils/types.py:320]: ``` :meth:`__post_init__` ```

Comment (gertjanvanzwieten) [reply nutils/types.py]: fixed

File: nutils/evaluable.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/evaluable.py]: I added an `abstract_property` descriptor class in `_util` that gets hold of the attribute name via `__set_name__`.

File: nutils/types.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/types.py]: correction: sphinx didn't like the change so I changed it to double backticks instead

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**